### PR TITLE
Duplicate contract code

### LINKED ISSUE #58: Add function App::duplicate_code
Add function `duplicate_code` to `App`.

### PR DESCRIPTION
**Pull request by DariuszDepta**
closes #58

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -21,7 +21,7 @@ pub enum Error {
     #[error("Unsupported wasm message: {0:?}")]
     UnsupportedWasmMsg(WasmMsg),
 
-    #[error("Unregistered code id")]
+    #[error("Unregistered code id: {0}")]
```
[the remark this answers is not in the document]
Comment (DariuszDepta) [reply src/error.rs:24]: I have checked on running `wasmd`` node (just to have this info for future use):
- when requested code id == 0 then the error message is:
**code id: invalid**
- when requested code id == [1..u64::MAX] and such code was not stored yet, then the error message is:
**code id {0}: no such code**
- when the code id is not an unsigned 64-bit integer, then the error message is from the parser but this is not our case.

So I guess, to be aligned with real blockchain functionality we should return two different error messages, first when the code id is zero and second when code id > 0 but there was no such code stored on blockchain.

I will add a new error code/message and adjust the existing one.

File: src/wasm.rs [hunk 2]
```diff
@@ -234,11 +241,46 @@ impl<ExecC, QueryC> WasmKeeper<ExecC, QueryC> {
...
+
+    /// Returns a handler to code of the contract with specified code id.
+    pub fn get_contract_code(&self, code_id: u64) -> AnyResult<&dyn Contract<ExecC, QueryC>> {
+        if code_id < 1 {
+            bail!(Error::UnregisteredCodeId(code_id));
+        }
+        let code_data = self
+            .code_data
+            .get((code_id - 1) as usize)
+            .ok_or(Error::UnregisteredCodeId(code_id))?;
```
[the remark this answers is not in the document]
Comment (DariuszDepta) [reply src/wasm.rs]: Good point! I have refactored it also.

File: src/error.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (DariuszDepta) [reply src/error.rs]: Changes already applied.

File: src/app.rs [hunk 3]
```diff
@@ -728,6 +728,28 @@ where
...
+    /// use cw_multi_test::App;
+    ///
+    /// let mut app = App::default();
+    ///
+    /// // there is no contract code with identifier 100 stored yet, returns an error
+    /// assert_eq!("Unregistered code id: 100", app.duplicate_code(100).unwrap_err().to_string());
+    ///
+    /// // zero is an invalid identifier for contract code, returns an error
+    /// assert_eq!("Unregistered code id: 0", app.duplicate_code(0).unwrap_err().to_string());
+    ///
```
[the remark this answers is not in the document]
Comment (DariuszDepta) [reply src/app.rs]: Adding the happy path as a documentation test requires some more refactoring. For now, I have added TODO comment and I add it as soon as the refactoring is done.

File: src/app.rs [hunk 4]
```diff
@@ -729,26 +729,57 @@ where
...
-    /// //   Provide a happy path example when testing contracts are public,
-    /// //   otherwise the example code would be unreadable.
+    /// // contract implementation
+    /// mod echo {
+    /// #  use cosmwasm_std::{Binary, Deps, DepsMut, Empty, Env, MessageInfo, Response, StdError, SubMsg, WasmMsg};
+    /// #  use serde::{Deserialize, Serialize};
+    /// #  use cw_multi_test::{Contract, ContractWrapper};
+    /// #
+    /// #  #[derive(Debug, Clone, Serialize, Deserialize, Default)]
+    /// #  pub struct EmptyMsg {}
```
[the remark this answers is not in the document]
Comment (DariuszDepta) [reply src/app.rs]: Oh, yes, thanks! I've used `Empty`.

File: src/app.rs [hunk 5]
```diff
@@ -729,26 +729,57 @@ where
...
+    /// #    let message = SubMsg::new(msg);
+    /// #    Ok(Response::new().add_submessage(message))
+    /// #  }
+    /// #
+    /// #  fn query(_deps: Deps, _env: Env, _msg: EmptyMsg) -> Result<Binary, StdError> {
+    /// #    Err(StdError::generic_err("not implemented yet"))
+    /// #  }
+    ///   // returns the contract
+    ///   pub fn contract() -> Box<dyn Contract<Empty>> {
+    /// #   Box::new(ContractWrapper::new(execute, instantiate, query))
```
[the remark this answers is not in the document]
Comment (DariuszDepta) [reply src/app.rs:762]: I have put the comment inside function body and added comment that the contract entry points are ommited in the example.

File: src/app.rs [hunk 6]
```diff
@@ -729,26 +729,57 @@ where
...
+    /// #  }
+    /// #
+    /// #  fn execute(_: DepsMut, _: Env, _info: MessageInfo, msg: WasmMsg) -> Result<Response, StdError> {
+    /// #    let message = SubMsg::new(msg);
+    /// #    Ok(Response::new().add_submessage(message))
+    /// #  }
+    /// #
+    /// #  fn query(_deps: Deps, _env: Env, _msg: EmptyMsg) -> Result<Binary, StdError> {
+    /// #    Err(StdError::generic_err("not implemented yet"))
+    /// #  }
```
[the remark this answers is not in the document]
Comment (DariuszDepta) [reply src/app.rs]: I've used `todo!()` macro, it works similarily but gives a hint that someone will implement it ;-)

**Review by chipshort — APPROVED** [review]
LGTM

**Pull request merged**